File: cli/po-cli/src/services/purchaseorderservice.ts
import { PurchaseOrderService as BasePurchaseOrderService, HttpService, Party } from 'wt-common-cli';

export class PurchaseOrderService extends BasePurchaseOrderService {
    constructor(httpService: HttpService) {
        super(httpService);
    }

    async createPurchaseOrder(buyer: Party, seller: Party, price: number, units: number, productGtin: number) {

        const purchaseOrderBody = {
            buyer,
            seller,
            price,
            units,
            productGtin
        };

        const partialOrder = await this.httpService.post('purchaseorders', purchaseOrderBody);

        return this.getPurchaseOrder(partialOrder.id);
    }

    async getPurchaseOrders(gln: number) {
        const purchaseOrders = await this.httpService.get(`purchaseorders?behalfOf=${gln}`);

        const purchaseOrderResponses = await this.getPurchaseOrderResponses(gln);

        for (const po of purchaseOrders) {
            for (const por of purchaseOrderResponses) {
                if (po.id === por.originalOrder.entityIdentification) {
                    po.response = por;
                    break;
                }
            }
        }

        return this.sortPurchaseOrders(purchaseOrders);
    }

    async getPurchaseOrder(poId: string) {
        return this.httpService.get(`purchaseorders/${poId}`);
    }

    async getPurchaseOrderResponse(poId: number) {
        return this.httpService.get(`purchaseorders/${poId}/responses`);
    }

    async getPurchaseOrderResponses(gln: number) {
        return this.httpService.get(`purchaseorders/responses?behalfOf=${gln}`);
    }

    async acceptPurchaseOrder(id: string) {
        return this.httpService.put(`purchaseorders/${id}/accept`, {});
    }

    async rejectPurchaseOrder(id: string) {
        return this.httpService.put(`purchaseorders/${id}/close`, {});
    }

    protected sortPurchaseOrders(purchaseOrders) {
        return purchaseOrders.sort((a, b) => {
            const aId = a.id.replace('PO', '');
            const bId = b.id.replace('PO', '');

            return aId - bId;
        });
    }
}

File: po-rest/src/main/java/org/tradefinance/main/services/PurchaseOrderService.java
package org.tradefinance.porest.services;

import java.util.Collection;

import org.tradefinance.assets.PurchaseOrder;
import org.tradefinance.assets.PurchaseOrderResponse;
import org.tradefinance.assets.defs.Party;

public interface PurchaseOrderService extends org.tradefinance.common.services.PurchaseOrderService {
    public Collection<PurchaseOrder> getPurchaseOrders(long behalfOfGln) throws Exception;

    public PurchaseOrder getPurchaseOrder(String id) throws Exception;

    public PurchaseOrder getPurchaseOrderByHash(String hash) throws Exception;

    public Collection<PurchaseOrderResponse> getPurchaseOrderResponses(long behalfOfGln) throws Exception;

    public PurchaseOrderResponse getPurchaseOrderResponse(String id) throws Exception;

    public PurchaseOrder createPurchaseOrder(Party buyer, Party seller, double price, int units, long productGtin) throws Exception;

    public void acceptPurchaseOrder(String id) throws Exception;

    public void closePurchaseOrder(String id) throws Exception;
}
